File: app/src/main/java/com/hover/stax/channels/ChannelDetailViewModel.java
package com.hover.stax.channels;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import com.hover.stax.database.DatabaseRepo;
import com.hover.stax.transactions.StaxTransaction;

import java.util.Calendar;
import java.util.List;

public class ChannelDetailViewModel extends AndroidViewModel {
	private static String TAG = "ChannelDetailViewModel";
	private DatabaseRepo repo;

	private MutableLiveData<Integer> id;
	private LiveData<Channel> channel;
	private LiveData<List<StaxTransaction>> transactions;
	private LiveData<Double> spentThisMonth;
	private LiveData<Double> feesThisYear;

	public ChannelDetailViewModel(@NonNull Application application) {
		super(application);
		repo = new DatabaseRepo(application);
		id = new MutableLiveData<>();
		channel = Transformations.switchMap(id, id -> repo.getLiveChannel(id));
		transactions =  Transformations.switchMap(id, id -> repo.getCompleteTransferTransactions(id));
		spentThisMonth = Transformations.switchMap(id, this::loadSpentThisMonth);
		feesThisYear = Transformations.switchMap(id, this::loadFeesThisYear);
	}

	void setChannel(int channelId) {
		new Thread(() -> {
			id.postValue(channelId);
		}).start();
	}

	LiveData<Double> loadSpentThisMonth(int id) {
		Calendar c = Calendar.getInstance();
		return repo.getSpentAmount(id, c.get(Calendar.MONTH) + 1, c.get(Calendar.YEAR));
	}

	LiveData<Double> loadFeesThisYear(int id) {
		Calendar c = Calendar.getInstance();
		return repo.getFees(id, c.get(Calendar.YEAR));
	}

	LiveData<Channel> getChannel() {
		if (channel == null) {
			channel = new MutableLiveData<>();
		}
		return channel;
	}

	LiveData<List<StaxTransaction>> getStaxTransactions() {
		if (transactions == null) {
			transactions = new MutableLiveData<>();
		}
		return transactions;
	}

	LiveData<Double> getSpentThisMonth() {
		if (spentThisMonth == null) {
			spentThisMonth = new MutableLiveData<>();
		}
		return spentThisMonth;
	}

	LiveData<Double> getFeesThisYear() {
		if (feesThisYear == null) { feesThisYear = new MutableLiveData<>(); }
		return feesThisYear;
	}
}

File: app/src/main/java/com/hover/stax/channels/ChannelDetailViewModel.kt
package com.hover.stax.channels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import com.hover.sdk.actions.HoverAction
import com.hover.stax.database.DatabaseRepo
import com.hover.stax.transactions.StaxTransaction
import java.util.*


class ChannelDetailViewModel(val repo: DatabaseRepo) : ViewModel() {

    private val id = MutableLiveData<Int>()
    var channel: LiveData<Channel> = MutableLiveData()
    var transactions: LiveData<List<StaxTransaction>> = MutableLiveData()
    var actions: LiveData<List<HoverAction>> = MutableLiveData()
    var spentThisMonth: LiveData<Double> = MutableLiveData()
    var feesThisYear: LiveData<Double> = MutableLiveData()

    private val calendar = Calendar.getInstance()

    init {
        channel = Transformations.switchMap(id, repo::getLiveChannel)
        transactions = Transformations.switchMap(id, repo::getChannelTransactions)
        actions = Transformations.switchMap(id, repo::getChannelActions)
        spentThisMonth = Transformations.switchMap(id, this::loadSpentThisMonth)
        feesThisYear = Transformations.switchMap(id, this::loadFeesThisYear)
    }

    fun setChannel(channelId: Int) = id.postValue(channelId)

    private fun loadSpentThisMonth(id: Int): LiveData<Double>? = repo.getSpentAmount(id, calendar.get(Calendar.MONTH) + 1, calendar.get(Calendar.YEAR))

    private fun loadFeesThisYear(id: Int): LiveData<Double>? = repo.getFees(id, calendar.get(Calendar.YEAR))
}